File: leetcode/sum/1. Two Sum/js/main.js
// Time O(N)
// Space O(N)
const twoSum = (nums, target) => {
  let map = new Map();
  let ans = [];

  for (let i = 0; i < nums.length; i++) {
    map.set(nums[i], i);
  }

  for (let i = 0; i < nums.length; i++) {
    let prev = target - nums[i];

    if (map.has(prev) && map.get(prev) !== i) {
      ans.push(i, map.get(prev));
      break;
    }
  }

  return ans;
};

File: leetcode/sum/1. Two Sum/py/main.py
# Time O(N)
# Space O(N)
class Solution:
    def twoSum(self, nums, target):
        memo = {}
        
        for i in range(len(nums)):
            x = target - nums[i]
            
            if x in memo:
                return [memo[x], i]
            
            memo[nums[i]] = i
            
        return []